diff --git a/Residual-Policy-Learning/controllers/robosuite/robosuiteNutAssemblyEnv.py b/Residual-Policy-Learning/controllers/robosuite/robosuiteNutAssemblyEnv.py
index 8d63c3f..0e95116 100644
--- a/Residual-Policy-Learning/controllers/robosuite/robosuiteNutAssemblyEnv.py
+++ b/Residual-Policy-Learning/controllers/robosuite/robosuiteNutAssemblyEnv.py
@@ -57,7 +57,7 @@ class NutAssembly(gym.Env):
         ob, reward, done, info = self.env.step(action)
         ob = self.env.observation_spec()
         observation = {}
-        observation['observation'] = np.hstack((ob['robot0_eef_pos'], ob['robot0_eef_quat'], ob['RoundNut0_pos'], ob['RoundNut0_quat']))
+        observation['observation'] = np.hstack((ob['robot0_eef_pos'], ob['robot0_eef_quat'], ob['RoundNut_pos'], ob['RoundNut_quat']))
         observation['desired_goal'] = np.array(self.env.sim.data.body_xpos[self.env.peg2_body_id])
         observation['achieved_goal'] = ob['RoundNut_pos']
         info['is_success'] = reward
@@ -67,7 +67,7 @@ class NutAssembly(gym.Env):
         ob = self.env.reset()
         ob = self.env.observation_spec()
         observation = {}
-        observation['observation'] = np.hstack((ob['robot0_eef_pos'], ob['robot0_eef_quat'], ob['RoundNut0_pos'], ob['RoundNut0_quat']))
+        observation['observation'] = np.hstack((ob['robot0_eef_pos'], ob['robot0_eef_quat'], ob['RoundNut_pos'], ob['RoundNut_quat']))
         observation['desired_goal'] = np.array(self.env.sim.data.body_xpos[self.env.peg2_body_id])
         observation['achieved_goal'] = ob['RoundNut_pos']
         return observation
diff --git a/requirements.txt b/requirements.txt
index 700015d..d6e1198 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,3 +1 @@
-cffi==1.14.0
-PyOpenGL==3.1.5
-pytest==6.1.2
+-e .
diff --git a/setup.py b/setup.py
index 109baa6..f1e4f2c 100644
--- a/setup.py
+++ b/setup.py
@@ -23,4 +23,11 @@ setup(
     eager_resources=["*"],
     include_package_data=True,
     python_requires=">=3",
+    description="robosuite: A Modular Simulation Framework and Benchmark for Robot Learning",
+    author="Yuke Zhu, Josiah Wong, Ajay Mandlekar, Roberto Martín-Martín",
+    url="https://github.com/ARISE-Initiative/robosuite",
+    author_email="yukez@cs.utexas.edu",
+    version="1.3.2",
+    long_description=long_description,
+    long_description_content_type="text/markdown",
 )
